### PR TITLE
Swift: extract types for patterns

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

File: swift/ql/lib/upgrades/147e087e57e51b2eb41e75c9c97380d0e6c20ecb/swift.dbscheme.stats [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<dbstats>
+     <typesizes />
+     <stats />
+</dbstats>
```
Comment (AlexDenisov) [inline swift/ql/lib/upgrades/147e087e57e51b2eb41e75c9c97380d0e6c20ecb/swift.dbscheme.stats:4]: We didn't put stats in any other upgrade scripts, do we need this now?

File: swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class Element extends @element {
+    string toString() {
+        none()
+    }
+}
+
+from Element p, Element t
+where none()
+select p, t
```
Comment (AlexDenisov) [inline swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql:9]: What does this script do? Is it simple removing all the elements?

**Review by AlexDenisov — COMMENTED** [review]
I'm not sure about the upgrade script, but overall LGTM.
Happy to approve once the formatting is fixed.

File: swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql (hunk 2 above)
Comment (geoffw0) [reply swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql:9]: I think `delete` in the `upgrade.properties` does this, see https://github.com/github/codeql/blob/main/docs/prepare-db-upgrade.md .

Comment (geoffw0) [reply swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql:9]: ... but that's what you've done in the downgrade script. Are we just trying to synthesize an empty table here?

**Review by geoffw0 — COMMENTED** [review]
Results look good as far as I can see.

It would be good to understand the motivation for this and whether this change meets that need.  I can certainly imagine these types being useful.

I also think we should have a DCA run on this.

File: swift/ql/lib/upgrades/147e087e57e51b2eb41e75c9c97380d0e6c20ecb/swift.dbscheme.stats (hunk 1 above)
Comment (MathiasVP) [reply swift/ql/lib/upgrades/147e087e57e51b2eb41e75c9c97380d0e6c20ecb/swift.dbscheme.stats:4]: I don't think we do, no.

**Comment by MathiasVP** [comment]
> It would be good to understand the motivation for this and whether this change meets that need. I can certainly imagine these types being useful.

The motivation for this is for type pruning, where we need to assign a meaningful type for dataflow nodes. Some time ago I talked to Robert about what we should do for the dataflow nodes representing patterns, and we agreed that we could probably assign it some notion of a Top type, but it looks like we might as well grab the real type off the patterns (which is what Robert is extracting here).

File: swift/ql/lib/change-notes/2023-10-24-pattern-types.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+category: minorAnalysis
+---
+
+* Tyes for patterns are now incluedd in the database and made available through the `Pattern::getType()` method.
```
Comment (geoffw0) [inline swift/ql/lib/change-notes/2023-10-24-pattern-types.md:5]:
```suggestion
* Tyes for patterns are now included in the database and made available through the `Pattern::getType()` method.
```

File: swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql (hunk 2 above)
Comment (geoffw0) [reply swift/ql/lib/upgrades/c0db61944f46ba5507f207ec2b1cff77ad0529a1/pattern_types.ql:9]: I'm not sure you need to.  I definitely don't consider myself an expert though.

**Comment by rdmarsh2** [comment]
DCA looks clean, only thing left is that question about the upgrade script - @AlexDenisov is there a better way to synthesize an empty table?

**Comment by geoffw0** [comment]
I *think* the `.dbscheme` defines what tables exist, the upgrade scripts only need to populate the right rows in them.  So to create an empty table no upgrade script is required.  Perhaps get confirmation from someone who works on upgrade scripts more often (maybe from the CPP side) or failing that the foundations team.

**Review by geoffw0 — COMMENTED** [review]
To be clear, the upgrade itself is still required (to mark the change in `.dbscheme`), I just don't think it needs to run any `.ql` script inside.  Terminology is confusing.

Other than that question, this LGTM as well.

**Review by geoffw0 — COMMENTED** [review]
There's still a spurious `swift.dbscheme.stats` in the upgrade script.

Assuming the upgrade / downgrade scripts still connect to the right versions, this otherwise LGTM.

**Comment by geoffw0** [comment]
I'm ready to approve this when the spurious `swift.dbscheme.stats` file is deleted.  Notably the upgrade script hash (`b83ff9c60c2bb4be2f3d1d4810268c557eb38f19`) is still current. :+1:

**Review by geoffw0 — APPROVED** [review]
:+1:

**Comment by rdmarsh2** [comment]
Lots of noise in the DCA analysis timings, but mostly faster and all with high standard deviations. I don't see any red flags.

**Pull request merged**